Cleanup

JingyiRose opened the pull request:
Enable commenting

aremath committed:
Major refactor of DTLOR DP

Makes the code more concise by applying reduce operations

aremath committed:
Incomplete improvements to DTLOR DP.

aremath committed:
First commit of the new (working) DP.

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -119,41 +126,34 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
         dynamic programming algorithm are explained in the tech report.
```
I think we are not currently returning the number of MPRs? Can you modify the comment to reflect the current outputs?

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -119,41 +126,34 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
             for eh in postorder(hostTree, "hTop"):
                 _,vh,eh1,eh2 = hostTree[eh]
                 eventsDict[(vp, vh, l_bottom)] = []
                 # is vh a tip?
```
delete this comment?

aremath committed:
Add Pipfile.

aremath committed:
Update python version for Pipfile.

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -78,7 +78,16 @@ def postorder(tree, rootEdgeName):
+
 # returns ce = oplus_i ce_i
+#TODO: rename
 def combine_costs(events_list):
```
one possible name I can think of is sth like find_min_cost_events

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -107,313 +126,221 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
-                    if switchList==[]:
-                        switchList=[[Infinity]]
-
+                    def get_transfer(l1, l2):
```
I think we only need one location here since the transferred child always has the same location as its parent? We can also discuss this more with Prof. Bush to see if this statement is actually well-supported by biological evidence.

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -107,313 +126,221 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
+                        # Cost to transfer ep2
+                        ep2_cost, ep2_locations = best_switch[(ep2, eh, l2)]
+                        ep2_switch_cost = T + synteny_cost + C[(ep1, eh, l1)][0] + ep2_cost
+                        ep2_switch_events = [("T", (ep1, vh, l1), (ep2, location[1], l2)) for location in \
```
the edges seem to be represented by their bottom vertex in the parasite and host tree, this can be a little bit confusing. I am not sure what's the best way to go about this

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -107,313 +126,221 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
-                if eh1 == None: # then eh2 == None too and vh is a tip!
-                    vhIsATip = True
+                # Is vh a tip?
+                if check_tip(vh, eh1, eh2):
```
since we are doing preorder here on species tree, we are probably checking whether vh is root, not tip~

JingyiRose commented on DTLOR_DP.py:
```diff
@@ -107,313 +126,221 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
+                    ep_best_switch = best_switch[(ep, eh, l_bottom)]
+                    O_eh2=O[(ep, eh2, l_bottom)]
+                    O_eh1=O[(ep, eh1, l_bottom)]
+                    best_switch[(ep, eh1, l_bottom)] = combine_costs([ep_best_switch, O_eh2, O_eh1])
```
I think best_switch is computed for this way:
best_switch[(ep, eh1)] = min(best_switch[(ep, eh)], O[(ep, eh2)])
best_switch[(ep, eh2)] = min(best_switch[(ep, eh)], O[(ep, eh1)])

aremath replied:
That's correct. Changed!

JingyiRose reviewed (commented):
The code looks really clean and is really easy to understand! Thanks for the great work Ross!
I've left a few comments we can discuss over the meeting~
btw, can you also add scipy = "*" to the Pipfile, thanks!

aremath committed:
Fix bugs in DTLOR_DP

 - Bug where the cost of an implicit origin event at the top of the gene
 tree was not being factored in.
 - Bug where best_switch was not being computed correctly

aremath committed:
More DTLOR Bugfixes

 - Fixed a bug where best_switch was being computed incorrectly
 - Compute transfers more quickly by fixing the syntenic location of the
 transferred child
 - Add scipy to Pipfile

aremath committed:
Fix bugs in DP

 - Fix bug in find_min_events where entries of C could be overwitten due
 to unintended mirroring

 - Fix bug in finding allsynteny where syntenic locations could be
 counted twice

aremath committed:
Removed pTop and hTop from the tree representation

Now the edge leading to the root of the tree is labeled with the name of
the root node. The trees are OrderedDicts where the root can be found by
using `next(iter(tree))`.

aremath committed:
Adding Bush's trees file to version control.
